### data from 4 H1 teratomas, generated by Daniella & processed by Yan before
### reprocessing using GRCh38_and_mm10 reference
### to save time, only processing the cells that are annotated as singlet by Yan's processing

# ----------------------------------------------------------------------
### preprocessing

## load cellranger output - sparse count matrix (human gene + mouse gene)
library(Seurat)
library(swne)
library(cowplot)
library(stringr)

WT <- list(Read10X('/media/Scratch_SSD_Voyager/sammi/4H1/1/dm-ter1/outs/filtered_feature_bc_matrix/'),
           Read10X('/media/Scratch_SSD_Voyager/sammi/4H1/2/dm-ter2/outs/filtered_feature_bc_matrix/'),
           Read10X('/media/Scratch_SSD_Voyager/sammi/4H1/3/dm-ter3/outs/filtered_feature_bc_matrix/'),
           Read10X('/media/Scratch_SSD_Voyager/sammi/4H1/4/dm-ter4/outs/filtered_feature_bc_matrix/'))
names(WT) <- c("WT1","WT2","WT3","WT4")

## process count matrix (remove mouse genes and omit 'GRCh38' from human genes before creating the seurat obj)
WT_cleanup.list <- list()
for (i in 1:length(WT)){
  genes.keep <- rownames(WT[i][[1]])[!grepl("^mm10", rownames(WT[i][[1]]))]
  WT_cleanup.list[i][[1]] <- WT[i][[1]][genes.keep,]
  
  row.names(WT_cleanup.list[i][[1]]) <- sapply(strsplit(row.names(WT_cleanup.list[i][[1]]), split = "GRCh38_"),"[[",2)
}
WT <- WT_cleanup.list
rm(WT_cleanup.list)

## create initial seuratobj
WT_SO.list <- list()
for (i in 1:length(WT)){
  WT_SO.list[i][[1]] <- CreateSeuratObject(counts = WT[[i]], project = "WT_H1",
                                               min.cells = round(0.001*ncol(WT[[i]])), min.features = 50)
}

## gene clean up (mito & ribo)
# count mito gene percent & remove mito gene
WT_SO.list.cleanup <- list()
for (i in 1:length(WT_SO.list)){
  WT_SO.list[i][[1]][["percent.mt"]] <- PercentageFeatureSet(WT_SO.list[i][[1]], pattern = "^MT-")
  
  genes.keep <- rownames(WT_SO.list[i][[1]])[!grepl("^MT-", rownames(WT_SO.list[i][[1]]))]
  WT_SO.list.cleanup[i][[1]] <- subset(WT_SO.list[i][[1]], features = genes.keep)
  
  counts.temp <- GetAssayData(WT_SO.list.cleanup[i][[1]], assay = "RNA")
  WT_SO.list.cleanup[i][[1]] <- CreateSeuratObject(counts.temp,
                                                       meta.data = WT_SO.list.cleanup[i][[1]]@meta.data)
  rm(counts.temp,genes.keep)
}
head(WT_SO.list.cleanup[1][[1]]@meta.data)
row.names(WT_SO.list.cleanup[1][[1]])
grep("^MT-", row.names(WT_SO.list.cleanup[1][[1]]))

WT_SO.list <- WT_SO.list.cleanup
rm(WT_SO.list.cleanup)

names(WT_SO.list) <- c("WT1","WT2","WT3","WT4")

## remove noncoding RNAs - RP11 is NOT in the gene names rn, decide to skip this removal for now unless ribo genes are driving


### subset to singlet previously defined
wt_counts <- "dm-ter-human-merged/"
wt_doublet <- "dm-ter-human-merged_doublets.txt"

counts_yw <- ReadData(wt_counts)
dim(counts_yw) #33354 30041
ncol(counts_yw)
doublets_yw <- as.logical(scan(wt_doublet, sep = "\n", what = numeric()))
doublets_yw[is.na(doublets_yw)] <- 1; sum(doublets_yw)/length(doublets_yw);
counts_yw <- counts_yw[,!doublets_yw]
dim(counts_yw) #33354 26744
tail(colnames(counts_yw))

# we have the merged matrix from Yan Wu, each cell annotated as a singlet has a colname of format "bc.ter",
# we'll extract these singlets from the new count matrix WT_SO.list (only has 4 nonbc - the first 4)

WT_SO.list.cleanup <- list()
for (i in 1:length(WT_SO.list)){
  WT_SO.list.cleanup[i][[1]] <- RenameCells(object = WT_SO.list[i][[1]], new.names = paste0(sapply(strsplit(colnames(WT_SO.list[i][[1]]), split = "-"),"[[",1),".",i))
}
colnames(WT_SO.list.cleanup[2][[1]])

length(which(sapply(strsplit(colnames(counts_yw), split = "[.]"),"[[",2) == "1")) #2353
length(intersect(colnames(counts_yw), colnames(WT_SO.list.cleanup[1][[1]]))) #2353
length(which(sapply(strsplit(colnames(counts_yw), split = "[.]"),"[[",2) == "2")) #1926
length(intersect(colnames(counts_yw), colnames(WT_SO.list.cleanup[2][[1]]))) #1926
length(which(sapply(strsplit(colnames(counts_yw), split = "[.]"),"[[",2) == "3")) #1487
length(intersect(colnames(counts_yw), colnames(WT_SO.list.cleanup[3][[1]]))) #1487
length(which(sapply(strsplit(colnames(counts_yw), split = "[.]"),"[[",2) == "4")) #5785
length(intersect(colnames(counts_yw), colnames(WT_SO.list.cleanup[4][[1]]))) #5681 - about 100 cells short

for (i in 1:length(WT_SO.list.cleanup)){
  WT_SO.list.cleanup[i][[1]] <- WT_SO.list.cleanup[i][[1]][,colnames(WT_SO.list.cleanup[i][[1]]) %in% intersect(colnames(counts_yw), colnames(WT_SO.list.cleanup[i][[1]]))]
}


# check these singlet to be classified as human cells by cellranger
WT_gem_cla <- list(read.table('/media/Scratch_SSD_Voyager/sammi/4H1/1/dm-ter1/outs/analysis/gem_classification.csv', sep = ",", header = T),
                   read.table('/media/Scratch_SSD_Voyager/sammi/4H1/2/dm-ter2/outs/analysis/gem_classification.csv', sep = ",", header = T),
                   read.table('/media/Scratch_SSD_Voyager/sammi/4H1/3/dm-ter3/outs/analysis/gem_classification.csv', sep = ",", header = T),
                   read.table('/media/Scratch_SSD_Voyager/sammi/4H1/4/dm-ter4/outs/analysis/gem_classification.csv', sep = ",", header = T))

length(WT_gem_cla)
head(WT_gem_cla[[1]],6)
sapply(WT_gem_cla,dim)

for (i in 1:length(WT_gem_cla)){
  WT_gem_cla[[i]]$barcode <- paste0(sapply(strsplit(WT_gem_cla[[i]]$barcode, split = "-"),"[[",1),".",i)
}

for (i in 1:length(WT_gem_cla)){
  WT_gem_cla[[i]] <- WT_gem_cla[[i]][WT_gem_cla[[i]]$barcode %in% colnames(WT_SO.list.cleanup[i][[1]]),]
}
sapply(WT_gem_cla,dim)

for (i in 1:length(WT_gem_cla)){
  print(paste0("teratoma ",i," mouse: ",length(which(WT_gem_cla[[i]]$call == "mm10"))))
  print(paste0("teratoma ",i," multiplet: ",length(which(WT_gem_cla[[i]]$call == "Multiplet"))))
}

# no mouse cell is detected but there are 41/48/21/288 multiplets, exclude them?

## put back all the metadata previously obtained

WT_meta <- read.table('dm-ter-human-merged_metadata.tsv', sep = "\t", header = T)
WT_meta <- WT_meta[WT_meta$barcoding == "nonbc",]

for (i in 1:length(WT_SO.list.cleanup)){
  meta.temp <- WT_meta$cluster
  names(meta.temp) <- WT_meta$cell
  WT_SO.list.cleanup[i][[1]] <- AddMetaData(WT_SO.list.cleanup[i][[1]], metadata = meta.temp, col.name = "cluster")
  
  meta.temp <- WT_meta$teratoma
  names(meta.temp) <- WT_meta$cell
  WT_SO.list.cleanup[i][[1]] <- AddMetaData(WT_SO.list.cleanup[i][[1]], metadata = meta.temp, col.name = "teratoma")
}
head(WT_SO.list.cleanup[1][[1]]@meta.data)

WT_SO.list <- WT_SO.list.cleanup
rm(WT_SO.list.cleanup)
names(WT_SO.list) <- c("WT1","WT2","WT3","WT4")

### before moving to seurat, check a couple things
for (i in 1:length(WT_SO.list)){
  # no cells have mouse genes
  print(length(grep("^mm10-", row.names(WT_SO.list[1][[1]]))))
  
  # no cells have mito genes
  print(length(grep("^MT-", row.names(WT_SO.list[1][[1]]))))
  
  # no cells are of mito per > 10
  print(length(which(WT_SO.list[i][[1]]$percent.mt > 10)))
  
  # check # of cells in each library
  print(dim(WT_SO.list[i][[1]])[2])
}

# a couple high mito content cells should be filtered out here
# save current seurat object list & then start seurat (filtering out these cells)

save(WT_SO.list, file = "Figure1_WildtypeH1_R1/WT4H1_YWsubset_preprocessing_mitoUNfiltered_20230102.Robj")

### seurat
WT_SO.list.orig <- WT_SO.list

# save violin plots pre and post filtering
p <- list(list())
for (i in 1:length(WT_SO.list)){
  p[i][[1]] <- plot_grid(VlnPlot(WT_SO.list[i][[1]], features = c("nFeature_RNA")) + NoLegend(),
                         VlnPlot(WT_SO.list[i][[1]], features = c("nCount_RNA")) + NoLegend(),
                         VlnPlot(WT_SO.list[i][[1]], features = c("percent.mt")) + NoLegend(),
                         ncol = 3, labels = names(WT_SO.list)[i], scale = 0.9)
}
pdf("Figure1_WildtypeH1_R1/Teratoma_4H1_Vln_postcellranger_prefiltered_20230102.pdf", width = 11, height = 6)
p
dev.off()

p <- list(list())
for (i in 1:length(WT_SO.list)){
  WT_SO.list[i][[1]] <- subset(WT_SO.list[i][[1]], subset = percent.mt < 10) #& nFeature_RNA > 200 & nFeature_RNA < 4000) 
  p[i][[1]] <- plot_grid(VlnPlot(WT_SO.list[i][[1]], features = c("nFeature_RNA")) + NoLegend(),
                         VlnPlot(WT_SO.list[i][[1]], features = c("nCount_RNA")) + NoLegend(),
                         VlnPlot(WT_SO.list[i][[1]], features = c("percent.mt")) + NoLegend(),
                         ncol = 3, labels = names(WT_SO.list)[i], scale = 0.9)
}
pdf("Figure1_WildtypeH1_R1/Teratoma_4H1_Vln_postcellranger_postfiltered_20230102.pdf", width = 11, height = 6)
p
dev.off()

sapply(WT_SO.list, dim)


# normalize & identify variable features for integration (4k variable genes)
for(i in 1:length(WT_SO.list)) {
  WT_SO.list[[i]] <- NormalizeData(object = WT_SO.list[[i]], verbose = FALSE)
  WT_SO.list[[i]] <- FindVariableFeatures(object = WT_SO.list[[i]], selection.method = "vst", 
                                        nfeatures = 4000, verbose = F)
}

## Identify anchors and integrate (which replaces runMultiCCA)
obj.anchors <- FindIntegrationAnchors(object.list = WT_SO.list, anchor.features = 4000, dims = 1:30)
obj.integrated <- IntegrateData(anchorset = obj.anchors, dims = 1:30)

## Switch to integrated assay. The variable features of this assay are 
## automatically set during IntegrateData
DefaultAssay(object = obj.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
obj.integrated <- ScaleData(object = obj.integrated, features = rownames(GetAssayData(obj.integrated)))
obj.integrated <- RunPCA(object = obj.integrated, npcs = 30, verbose = F)
obj.integrated <- RunUMAP(object = obj.integrated, reduction = "pca", 
                          dims = 1:30, metric = "cosine")

plot_grid(DimPlot(obj.integrated, group.by = "cluster", label = T, repel = T) + NoLegend(),
DimPlot(obj.integrated, group.by = "teratoma", label = F, repel = T) )

t(table(obj.integrated$teratoma, obj.integrated$cluster))

### merge some cell types for more read-out in RNA editing

obj.integrated$annot_v1 <- obj.integrated$cluster
#Neural Progenitors: RG + CycProg
obj.integrated$annot_v1[obj.integrated$cluster %in% c("Radial Glia","CycProg")] <- "Neural-Progenitors"
#Neurons: Early Neurons + Retinal Neurons
obj.integrated$annot_v1[obj.integrated$cluster %in% c("Early Neurons","Retinal Neurons")] <- "Neurons"
#SCP: Schwann + Melanobalsts
obj.integrated$annot_v1[obj.integrated$cluster %in% c("Schwann Cells","Melanoblasts")] <- "SCP"

#Gut Epithelial: Foregut Epi + Mid/Hindgut Epi + Airway Epi
obj.integrated$annot_v1[obj.integrated$cluster %in% c("Airway Epi","Foregut Epi","Mid/Hindgut Epi")] <- "Gut-Epi"

#Hematopoietic: HSC + immune + Erythrocyte
obj.integrated$annot_v1[obj.integrated$cluster %in% c("HSC","Immune","Erythrocyte")] <- "Hematopoietic"
#Muscle: Muscle Prog + Cardiac/Skeletal Muscle
obj.integrated$annot_v1[obj.integrated$cluster %in% c("Muscle Prog","Cardiac/Skeletal Muscle")] <- "Muscle"

table(obj.integrated$annot_v1)
table(obj.integrated$annot_v1, obj.integrated$teratoma)

### fix cell type names
obj.integrated$annot_v1 <- str_replace_all(string = obj.integrated$annot_v1, pattern = "/", replacement = "-")
obj.integrated$annot_v1 <- str_replace_all(string = obj.integrated$annot_v1, pattern = fixed(" "), replacement = "-")
table(obj.integrated$annot_v1)

# ----------------------------------------------------------------------

# save seurat obj
save(obj.integrated, file = "Figure1_WildtypeH1_R1/WT4H1_YWsubset_integratedObj_20230102.Robj")

load("Figure1_WildtypeH1_R1/WT4H1_YWsubset_integratedObj_20230102.Robj")

obj.integrated <- RunUMAP(object = obj.integrated, reduction = "pca", 
                          dims = 1:30, metric = "cosine")
WT_4H1 <- obj.integrated
save(WT_4H1, file = "Figure1_WildtypeH1_R1/WT4H1_YWsubset_integratedObj_umapmodel_20230115.Robj")

load("Figure1_WildtypeH1_R1/WT4H1_YWsubset_integratedObj_umapmodel_20230115.Robj")

cluster_levels <- c("Radial Glia","CycProg","Early Neurons","Retinal Neurons","Retinal Epi","Schwann Cells","Melanoblasts",
                    "Foregut Epi","Mid/Hindgut Epi","Airway Epi","HSC","Immune","Erythrocyte",
                    "Adipogenic MSC/Fib","Chondrogenic MSC/Fib","Cycling MSC/Fib","MSC/Fib","MyoFib",
                    "Muscle Prog","Cardiac/Skeletal Muscle","Pericytes","Smooth Muscle",
                    "Kidney Prog")

obj.integrated$cluster <- factor(obj.integrated$cluster, levels = cluster_levels)

#90BFF9 = #4795F5 + #207FF3
#C0C0FF = #7474FF + #DADAFF
#00FF00 = #9EEA00 + #4FF4A2
#005A00 = #002700 + #005A00 + #00A700
#000080 = #000032 + #04409B + #4545FF
#FF8000 = #B35A00 + #FF9933


cluster_col <- c("#4795F5","#0B62CD","#7474FF","#DADAFF","#F2B77C","#9EEA00","#4FF4A2",
                 "#002700","#005A00","#00A700","#000032","#04409B","#4545FF",
                 "#0000FF","#FF0000","#00C000","#AD07E3","#000000",
                 "#B35A00","#FF9933","#A00000","#94641F","#610051")
names(cluster_col) <- cluster_levels

pdf("Figure1_WildtypeH1_R1/WT4H1_umap_by_cluster_20230201.pdf")
DimPlot(obj.integrated, group.by = "cluster", label = T, repel = T, cols = cluster_col, shuffle = T) + NoLegend()
dev.off()

# ----------------------------------------------------------------------
### plot correlation to transcriptomics of cells in Jay Shendure's fetal cell atlas
load("WT4H1_YWsubset_integratedObj_20230102.Robj")
# gene exp validation by Jay Shendure's fetal cell atlas
load("/media/Home_Raid1_Voyager/sammi/Teratoma_Analysis/TimeSeries/JSHF_SO_down100k.Robj")
JSHF_ct_collapse <- read.table("/media/Scratch_SSD_Voyager/sammi/20220406_Teratoma_TS/JSHF_Collapsing.csv", header = TRUE, sep = ",")

levels(factor(JSHF_SO_down$Main_cluster_name))
list_cluster <- factor(JSHF_SO_down$Main_cluster_name)
cl.replace <- JSHF_ct_collapse$New.Ident
names(cl.replace) <- JSHF_ct_collapse$Orig.Ident
list_cluster <- plyr::revalue(list_cluster, replace = cl.replace)
JSHF_SO_down$Collapsed_CT <- list_cluster
table(JSHF_SO_down$Collapsed_CT)
JSHF_ST_co_order <- c("Astrocytes","Neurons","Ex.Neurons","In.Neurons","In.Interneurons",
                      "Retinal.Prog","Retinal.Neurons","Retinal.Interneurons",
                      "Ganglion.Cells","Horizontal.Cells","Retinal.Pig",
                      "Retinal.Epi","Retinal.Fiber","Oligodendrocytes",
                      "Neuroendocrine.Cells",
                      "Schwann.Cells","Glia",
                      
                      "Foregut.Epi","Lung.Epi","Gut.Epi","Heart.Epi","Heart.Cells",
                      "Lymphatic.Endo","Lymphatic.Epi","Endometrial.Epi","Vascular.Endo",
                      
                      "Microglia","Immune.Cells","Blood.Cells",
                      "MSC","Stromal.Cells","Skeletal.Muscle.Cells", "Cardiac.Muscle.Cells",
                      "Mesothelial.Cells","Smooth.Muscle.Cells",
                      "Kidney",
                      
                      "Squamous.Epithelial.Cells","Hepatoblasts","Sympathoblasts","Trophoblasts")
JSHF_SO_down$Collapsed_CT <- factor(JSHF_SO_down$Collapsed_CT, levels = JSHF_ST_co_order)
length(levels(factor(JSHF_SO_down$Collapsed_CT)))

hist(JSHF_SO_down@assays$RNA@counts@x)
JSHF_SO_down <- NormalizeData(JSHF_SO_down)
JSHF_SO_down <- FindVariableFeatures(JSHF_SO_down, nfeatures = 3000)
JSHF_SO_down <- ScaleData(JSHF_SO_down, features = VariableFeatures(JSHF_SO_down))
genes.use <- unique(c(as.character(YanMarkersPub), VariableFeatures(JSHF_SO_down), VariableFeatures(obj.integrated)))
genes.use <- Reduce(intersect, list(genes.use, rownames(GetAssayData(JSHF_SO_down, assay = "RNA", slot = "scale.data")), 
                                    rownames(GetAssayData(obj.integrated, assay = "integrated"))))
length(genes.use) # 970
Tera.r.assay <- GetAssayData(obj.integrated, assay = "integrated")

cor.result <- MapClustersCor(query.data = Tera.r.assay, query.clusters = obj.integrated$cluster, 
                             train.data = GetAssayData(JSHF_SO_down, assay = "RNA", slot = "scale.data"),
                             train.clusters = JSHF_SO_down$Collapsed_CT,
                             genes.use = genes.use)
ggHeat(t(cor.result$cluster.correlations[complete.cases(cor.result$cluster.correlations),]), clustering = "both", x.lab.size = 11, y.lab.size = 11)
# when plotting, remove neuroendocrine cells, blood cells, heart epi, heart cells, lymphatic endo, lymphatic epi, endometrial epi
# mesothelial cells, squamous epithelial cells, hepatoblasts, sympathoblasts, trophoblasts
ggHeat(t(cor.result$cluster.correlations[complete.cases(cor.result$cluster.correlations),][JSHF_ST_co_order[-c(15,21,22,23,24,25,29,34,37,38,39,40)],]), x.lab.size = 11, y.lab.size = 11)

pdf("Figure1_WildtypeH1_R1/WT_4ter_correlateJSHF_20230221.pdf", height = 7, width = 10)
ggHeat(t(cor.result$cluster.correlations[complete.cases(cor.result$cluster.correlations),][JSHF_ST_co_order[-c(15,21,22,23,24,25,29,34,37,38,39,40)],cluster_levels]), x.lab.size = 11, y.lab.size = 11) + 
  labs(title='Correlation between Cell Types from Teratoma and Fetal Cell Atlas', x='Fetal Cell Atlas Cell Types', y='Teratoma Time Series Cell Types')
dev.off()

YanMarkersSub <- intersect(c("VIM","SOX2","HES5","HMGB2","RORB","CLVS1","CRB1","DCX","MAP2",
                             "OTX2","NRL","MITF","TTR",
                             "MPZ","SOX10","MLANA",
                             "SOX17","FOXA2","ELF3","PAX9","KRT4","FOXJ1","CDHR3","CDX2","RBPJL","GP2",
                             "CD74","CD34","HHEX","THY1","ITM2A","SHOX2","COL2A1","SOX9","COL15A1",
                             "MYOD1","PAX7","TNNT1","TNNI2","FOXC1","CYP1B1","ACTA2","RGS5","WT1"#,"FMN1","PKHD1","PAX2","PAX8","EYA1"
), VariableFeatures(obj.integrated, assay = "integrated"))
heat.mat <- t(apply(GetAssayData(obj.integrated, assay = "integrated", slot = "scale.data")[YanMarkersSub,], 1, function(x) {
  tapply(x, obj.integrated$cluster, mean)
}))
heat.mat[heat.mat > 5] <- 5

pdf("Figure1_WildtypeH1_R1/WT_4ter_MarkerExpression_20230221.pdf", width = 6, height = 11)
ggHeat(heat.mat[,cluster_levels], clustering = "none", x.lab.size = 11, y.lab.size = 11) + ggtitle('Marker Gene Expression for Wild Type Teratoma Cell Types')
dev.off()

# ------------------------------------------------------------
# further split up neuroectoderm
load("WT4H1_YWsubset_integratedObj_20230102.Robj")
obj.integrated
Idents(obj.integrated) <- obj.integrated$cluster

obj.integrated.neu <- subset(obj.integrated, ident = c("CycProg","Early Neurons","Radial Glia","Retinal Epi","Retinal Neurons","Schwann Cells","Melanoblasts"))

DefaultAssay(obj.integrated.neu) <- "RNA"
manu_markers <- c("GAD2", "DLX1", "DLX5",                        #Gabaergic
                  "NEUROD2","MAPT","SNAP25","GRIA2",             #Glutamatergic
                  "TTYH1","MT2A","SLC1A3",                       #Glia
                  "GADD45G","EOMES","TAC3","NHLH1",              #Intermediate Progenitors
                  "ID3","ID1","PAX6","CA2","OTX2")               #Progenitor
                  #"AIF1","TMEM119","SALL1","PTPRC","CD68","ITGAM"
pdf('Figure1_WildtypeH1_R1/WT_4ter_corticalorganoidMarkerExpression_dotplot_20230718.pdf', height = 6, width = 8)
DotPlot(obj.integrated.neu, features = manu_markers) + RotatedAxis()
dev.off()

#pdf('Figure1_WildtypeH1_R1/WT_4ter_corticalorganoidMarkerExpression_featureplot_20230718.pdf', height = 20, width = 20)
#FeaturePlot(obj.integrated.neu, features = manu_markers)
#dev.off()

DefaultAssay(obj.integrated.neu) <- "integrated"
obj.integrated.neu <- FindNeighbors(obj.integrated.neu, dims = 1:30)
obj.integrated.neu <- FindClusters(obj.integrated.neu, algorithm = 2, resolution = 2)

DimPlot(obj.integrated.neu, group.by = "seurat_clusters", label = T, repel = T)
table(obj.integrated.neu$seurat_clusters)
DefaultAssay(obj.integrated.neu) <- "RNA"
DotPlot(obj.integrated.neu, features = manu_markers, group.by = "seurat_clusters") + RotatedAxis()
DotPlot(obj.integrated.neu, features = c("SLC17A7","SLC17A6","GAD1","GAD2"), group.by = "seurat_clusters") + RotatedAxis()


obj.integrated.neu$annot_l2_v1 <- NA
for (i in 1:ncol(obj.integrated.neu)){
  if (obj.integrated.neu@meta.data[i, "cluster"] %in% c("CycProg","Radial Glia","Retinal Epi","Retinal Neurons","Schwann Cells","Melanoblasts")){
    obj.integrated.neu@meta.data[i,"annot_l2_v1"] <- obj.integrated.neu@meta.data[i, "cluster"]
  }
}
table(obj.integrated.neu$annot_l2_v1)


min_exp <- 0.25
for (i in 1:ncol(obj.integrated.neu)){
  if (is.na(obj.integrated.neu@meta.data[i,"annot_l2_v1"])) {
    if (log(obj.integrated.neu@assays$RNA@counts["EOMES",i]) > min_exp | log(obj.integrated.neu@assays$RNA@counts["GADD45G",i]) > min_exp | log(obj.integrated.neu@assays$RNA@counts["NHLH1",i]) > min_exp &
    log(obj.integrated.neu@assays$RNA@counts["SLC17A7",i]) < min_exp  & log(obj.integrated.neu@assays$RNA@counts["SLC17A6",i]) < min_exp &
    log(obj.integrated.neu@assays$RNA@counts["GAD2",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["GAD1",i]) < min_exp ){
        obj.integrated.neu@meta.data[i,"annot_l2_v1"] <- "Intermediate Progenitor"
    }
  }
}
table(obj.integrated.neu$annot_l2_v1)

for (i in 1:ncol(obj.integrated.neu)){
  if (is.na(obj.integrated.neu@meta.data[i,"annot_l2_v1"])) {
    if (log(obj.integrated.neu@assays$RNA@counts["SLC17A7",i]) > min_exp | log(obj.integrated.neu@assays$RNA@counts["SLC17A6",i]) > min_exp &
    log(obj.integrated.neu@assays$RNA@counts["GAD2",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["GAD1",i]) < min_exp &
    log(obj.integrated.neu@assays$RNA@counts["EOMES",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["GADD45G",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["NHLH1",i]) < min_exp){
        obj.integrated.neu@meta.data[i,"annot_l2_v1"] <- "Glutamatergic"
    }
  }
}
table(obj.integrated.neu$annot_l2_v1)

for (i in 1:ncol(obj.integrated.neu)){
  if (is.na(obj.integrated.neu@meta.data[i,"annot_l2_v1"])) {
    if (log(obj.integrated.neu@assays$RNA@counts["GAD2",i]) > min_exp | log(obj.integrated.neu@assays$RNA@counts["GAD1",i]) > min_exp  &
    log(obj.integrated.neu@assays$RNA@counts["SLC17A7",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["SLC17A6",i]) < min_exp  &
    log(obj.integrated.neu@assays$RNA@counts["EOMES",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["GADD45G",i]) < min_exp & log(obj.integrated.neu@assays$RNA@counts["NHLH1",i]) < min_exp){
        obj.integrated.neu@meta.data[i,"annot_l2_v1"] <- "GABAergic"
    }
  }
}
table(obj.integrated.neu$annot_l2_v1)

for (i in 1:ncol(obj.integrated.neu)){
  if (is.na(obj.integrated.neu@meta.data[i,"annot_l2_v1"])) {
        obj.integrated.neu@meta.data[i,"annot_l2_v1"] <- "Early Neurons"
  }
}
table(obj.integrated.neu$annot_l2_v1)

table(obj.integrated.neu$annot_l2_v1, obj.integrated.neu$teratoma)
DotPlot(obj.integrated.neu, features = c("EOMES","GADD45G","NHLH1","SLC17A7","SLC17A6","GAD1","GAD2","RBFOX3","NEUROD2"), group.by = "annot_l2_v1") + RotatedAxis()
head(obj.integrated.neu)
obj.integrated.neu$annot_l2_v1_ter <- paste0(obj.integrated.neu$annot_l2_v1, "_", obj.integrated.neu$teratoma)
DotPlot(obj.integrated.neu, features = c("EOMES","GADD45G","NHLH1","SLC17A7","SLC17A6","GAD1","GAD2","RBFOX3","NEUROD2"), group.by = "annot_l2_v1_ter") + RotatedAxis()
DotPlot(obj.integrated.neu, features = c("ADAR","ADARB1","ADARB2"), group.by = "annot_l2_v1_ter") + RotatedAxis()


Idents(obj.integrated.neu) <- obj.integrated.neu$annot_l2_v1
markers_oen <- FindMarkers(obj.integrated.neu, ident.1 = "Early Neurons")


neuro_ct_markers <- c("HMGB2","PCLAF","TOP2A","CENPF","MKI67",
                      "ATP1A2","ANGPT1",
                      "ID2","SOX2","NES","FABP7",
                      "NCAM1","MBP","BCAN","RFX4",
                      "FOXG1","LHX2",
                      "EOMES","NEUROG2",
                      "SLC17A7","NEUROD6","FEZF2",
                      "SLC17A6","KCNJ6","LMX1B",
                      "POU3F2","ASCL1","PRMT8",
                      "ASIC2","TAFA1","FSTL5","GRID1",
                      "DLX6-AS1","GAD2",
                      "SLC6A5","RBFOX1","GAD1","ADARB2","LAMP5",
                      "TFEC","YAP1",
                      "OTX2","MITF","PMEL","BEST1","RLBP1",
                      "SIX6","VSX2","RAX","HES1",
                      "TRPM1",
                      "PRDM13","ONECUT1",
                      "GAP43","POU4F2","PRPH",
                      "RSPO2","SOX2-OT","ZIC5",
                      "LMX1A","FOXJ1","PIFO",
                      "MPZ","SOX10"
                      )

DefaultAssay(obj.integrated.neu) <- "RNA"
pdf('Figure1_WildtypeH1_R1/WT_4ter_neuroctMarkerExpression_annot_l1_v1_dotplot_20230722.pdf', height = 6, width = 18)
DotPlot(obj.integrated.neu, features = neuro_ct_markers, group.by = "annot_l2_v1") + RotatedAxis()
dev.off()

# save some results
table(obj.integrated.neu$annot_l2_v1)
obj.integrated.neu$annot_l2_v1 <- factor(obj.integrated.neu$annot_l2_v1, levels = c('Radial Glia',"CycProg","Intermediate Progenitor","Glutamatergic","GABAergic","Early Neurons","Retinal Epi","Retinal Neurons","Schwann Cells","Melanoblasts"))
table(obj.integrated.neu$annot_l2_v1)

table(obj.integrated.neu$annot_l2_v1, obj.integrated.neu$teratoma)
write.table(table(obj.integrated.neu$annot_l2_v1, obj.integrated.neu$teratoma), file = "Figure1_WildtypeH1_R1/WT_4ter_neuroecto_cellcounttable.txt", quote = F, sep = ",")

save(obj.integrated.neu, file = "Figure1_WildtypeH1_R1/WT_4ter_neuroectoderm_20230722.rda")

obj.integrated.neu.list <- SplitObject(obj.integrated.neu, split.by = "teratoma")

for(i in 1:length(obj.integrated.neu.list)) {
  obj.integrated.neu.list[[i]] <- NormalizeData(object = obj.integrated.neu.list[[i]], verbose = FALSE)
  obj.integrated.neu.list[[i]] <- FindVariableFeatures(object = obj.integrated.neu.list[[i]], selection.method = "vst", 
                                        nfeatures = 4000, verbose = F)
}

obj.anchors <- FindIntegrationAnchors(object.list = obj.integrated.neu.list, anchor.features = 4000, dims = 1:30)
obj.integrated.neu <- IntegrateData(anchorset = obj.anchors, dims = 1:30)

#obj.integrated.neu <- ScaleData(object = obj.integrated.neu, features = VariableFeatures(obj.integrated.neu))
obj.integrated.neu <- ScaleData(object = obj.integrated.neu, features = rownames(GetAssayData(obj.integrated.neu)))
obj.integrated.neu <- RunPCA(object = obj.integrated.neu, npcs = 30, verbose = F)
obj.integrated.neu <- RunUMAP(object = obj.integrated.neu, reduction = "pca", 
                          dims = 1:30, metric = "cosine")

cluster_levels <- c('Radial Glia',"CycProg","Intermediate Progenitor","Glutamatergic","GABAergic","Early Neurons","Retinal Epi","Retinal Neurons","Schwann Cells","Melanoblasts")
cluster_col <- c("#B5D33D","#8931EF","#6CA2EA","#A62639","#EB7D5B","#FF00BD","#009F75","#004488","#B09334","#39CCCC")
DimPlot(obj.integrated.neu, group.by = "teratoma",cols = cluster_col, label = T, repel = T)


pdf('Figure1_WildtypeH1_R1/WT_4ter_neuroecto_umap_20230719.pdf',width=10,height=10)
DimPlot(obj.integrated.neu, group.by = "annot_l2_v1",cols = cluster_col, label = T, repel = T)
dev.off()


# generate an overall umap but with neuroectoderm cells colored differently

load("Figure1_WildtypeH1_R1/WT4H1_YWsubset_integratedObj_umapmodel_20230115.Robj")
load("Figure1_WildtypeH1_R1/WT_4ter_neuroectoderm_20230722.rda")

DefaultAssay(WT_4H1) <- "RNA"
FeaturePlot(WT_4H1, features = c("SLC17A6","GAD2"))

WT_4H1$annot_l2_v1 <- WT_4H1$cluster
for (i in colnames(obj.integrated.neu)){
  WT_4H1@meta.data[i,"annot_l2_v1"] <- as.character(obj.integrated.neu@meta.data[i,"annot_l2_v1"])
}
table(WT_4H1$annot_l2_v1)
WT_4H1$annot_l2_v1 <- factor(WT_4H1$annot_l2_v1, levels = cluster_levels)


cluster_levels <- c("Radial Glia","CycProg","Intermediate Progenitor","Glutamatergic","GABAergic","Early Neurons","Retinal Neurons","Retinal Epi","Schwann Cells","Melanoblasts",
                    "Foregut Epi","Mid/Hindgut Epi","Airway Epi","HSC","Immune","Erythrocyte",
                    "Adipogenic MSC/Fib","Chondrogenic MSC/Fib","Cycling MSC/Fib","MSC/Fib","MyoFib",
                    "Muscle Prog","Cardiac/Skeletal Muscle","Pericytes","Smooth Muscle",
                    "Kidney Prog")

cluster_col <- c("#4795F5","#0B62CD","#6CA2EA","#A62639","#EB7D5B","#7474FF","#DADAFF","#F2B77C","#9EEA00","#4FF4A2",
                 "#002700","#005A00","#00A700","#000032","#04409B","#4545FF",
                 "#0000FF","#FF0000","#00C000","#AD07E3","#000000",
                 "#B35A00","#FF9933","#A00000","#94641F","#610051")
names(cluster_col) <- cluster_levels

pdf('Figure1_WildtypeH1_R1/WT_4ter_allcelltypes_neuroectosubset_umap_20230722.pdf',width=10,height=10)
DimPlot(WT_4H1, group.by = "annot_l2_v1",cols = cluster_col, label = T, repel = T)
dev.off()

### save cell barcode list for pseudobulk
for (i in names(table(WT_4H1$annot_v1))){
  for (j in names(table(WT_4H1$teratoma))){
    if (length (row.names(WT_4H1@meta.data[WT_4H1$annot_v1 == i & WT_4H1$teratoma == j, ])) != 0){
      write.table(paste0("CB:Z:",sapply(strsplit(row.names(WT_4H1@meta.data[WT_4H1$annot_v1 == i & WT_4H1$teratoma == j, ]), split = "[.]"), "[[",1), "-1"), 
                  paste0('cb_listsH1/v4/',i,'_',j,'.txt'), col.names = FALSE, row.names = FALSE, 
                  quote = FALSE, sep = ',')
    }
  }
}

for (i in names(table(obj.integrated.neu$annot_l2_v1))){
  for (j in names(table(obj.integrated.neu$teratoma))){
    if (length (row.names(obj.integrated.neu@meta.data[obj.integrated.neu$annot_l2_v1 == i & obj.integrated.neu$teratoma == j, ])) != 0){
      write.table(paste0("CB:Z:",paste0(sapply(strsplit(row.names(obj.integrated.neu@meta.data[obj.integrated.neu$annot_l2_v1 == i & obj.integrated.neu$teratoma == j, ]), split = "[.]"), "[[",1)),"-1"), 
                  paste0('/media/Scratch_SSD_Voyager/sammi/RNA_editing/cb_listsH1/v5/',i,'_',j,'.txt'), col.names = FALSE, row.names = FALSE, 
                  quote = FALSE, sep = ',')
    }
  }
}